controllers/skipjob/job.go

```diff
@@ -0,0 +1,387 @@
+package skipjobcontroller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	skiperatorv1alpha1 "github.com/kartverket/skiperator/api/v1alpha1"
+	"github.com/kartverket/skiperator/pkg/resourcegenerator/core"
+	"github.com/kartverket/skiperator/pkg/resourcegenerator/gcp"
+	"github.com/kartverket/skiperator/pkg/util"
+	"golang.org/x/exp/maps"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"time"
+)
+
+var (
+	SKIPJobReferenceLabelKey = "skipJobOwnerName"
+	DefaultPollingRate       = time.Second * 15
+
+	IstioProxyPodContainerName = "istio-proxy"
+)
+
+func (r *SKIPJobReconciler) reconcileJob(ctx context.Context, skipJob *skiperatorv1alpha1.SKIPJob) (reconcile.Result, error) {
+	job := batchv1.Job{ObjectMeta: metav1.ObjectMeta{
+		Namespace: skipJob.Namespace,
+		Name:      skipJob.HashedName(),
+	}}
+
+	cronJob := batchv1.CronJob{ObjectMeta: metav1.ObjectMeta{
+		Namespace: skipJob.Namespace,
+		Name:      skipJob.HashedName(),
+	}}
+
+	gcpIdentityConfigMap, err := r.getGCPIdentityConfigMap(ctx, *skipJob)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if skipJob.Spec.Cron != nil {
+		err := r.DeleteObjectIfExists(ctx, &job)
```

Comment:
Was this the place we were supposed to use immutability instead of deleting the teams' jobs?

Reply:
Hmm.. might also be an idea to have immutability for switching from a cronjob to a normal job and vice versa. As of now this function, and the similar one in the case that `skipJob.Spec.Cron == nil`, simply delete what the opposite case would have created.

Not sure if we want to allow switching between these. I have not added immutability through kubebuilder, but rather I'm diffing the specific parts of a CronJob / Job we do not allow changes to. This might not be the best solution, as disallowing on admission control might be more obvious

Reply:
Yeah, I think that is a job for the admission controller

Reply:
I believe things should be fine now @esphen!

* If unset, CronSettings may not be added, and if set they may not be removed.
* ContainerSettings may not be changed
* If JobSettings/CronSettings are updated they are allowed to do so

However, I'm not entirely sure how to test this with kuttl, as the admission controller causes an error.. 🤔